src/headers/frame_header.rs

```diff
@@ -0,0 +1,544 @@
+        have_crop : have_crop, x0: x0, y0: y0, width: width, height: height,
+        img_width: nonserialized.img_width, img_height: nonserialized.img_height)]
+    blending_info: BlendingInfo,
+
+    #[size_coder(explicit(nonserialized.num_extra_channels))]
+    #[default_element(BlendingInfo::default())]
+    #[nonserialized(num_extra_channels : nonserialized.num_extra_channels,
+        have_crop : have_crop, x0: x0, y0: y0, width: width, height: height,
+        img_width: nonserialized.img_width, img_height: nonserialized.img_height)]
+    ec_blending_info: Vec<BlendingInfo>,
+
+    #[coder(u2S(0, 1, Bits(8), Bits(32)))]
+    #[default(0)]
+    #[condition((frame_type == FrameType::RegularFrame ||
+        frame_type == FrameType::SkipProgressive) && nonserialized.have_animation)]
+    duration: u32,
+
+    #[coder(Bits(32))]
+    #[default(0)]
+    #[condition((frame_type == FrameType::RegularFrame ||
+        frame_type == FrameType::SkipProgressive) && nonserialized.have_timecode)]
+    timecode: u32,
+
+    #[default(frame_type == FrameType::RegularFrame)]
+    #[condition(frame_type == FrameType::RegularFrame || frame_type == FrameType::SkipProgressive)]
+    pub is_last: bool,
+
+    #[coder(Bits(2))]
+    #[default(0)]
+    #[condition(frame_type != FrameType::LFFrame && !is_last)]
+    save_as_reference: u32,
+
+    // TODO(TomasKralCZ): figure out a way of extracting this huge default into separate variables
+    /* The default value for save_before_ct is d_sbct = !(full_frame && (frame_type ==
+    kRegularFrame || frame_type == kSkipProgressive) && blending_info.mode ==
+    kReplace && (duration == 0 || save_as_reference != 0) && !is_last). */
+    #[default(!(!have_crop || (x0 == 0 && y0 == 0 &&
+        width as i64 + x0 as i64 >= nonserialized.img_width as i64 &&
+        height as i64 + y0 as i64 >= nonserialized.img_height as i64))
+        && (frame_type == FrameType::RegularFrame || frame_type == FrameType::SkipProgressive) &&
+        blending_info.mode == BlendingMode::Replace && (duration == 0 || save_as_reference != 0) && !is_last)]
+    #[condition(frame_type != FrameType::LFFrame)]
+    save_before_ct: bool,
+
+    name: String,
+
+    #[default(RestorationFilter::default())]
+    #[nonserialized(encoding : encoding)]
+    restoration_filter: RestorationFilter,
+
+    #[default(Extensions::default())]
+    extensions: Extensions,
+}
+
+impl FrameHeader {
+    fn check(&self, nonserialized: &FrameHeaderNonserialized) -> Result<(), Error> {
+        if let Some(val) = self.jpeg_upsampling.iter().find(|x| **x > 3) {
+            return Err(Error::InvalidJpegUpsampling(*val));
+        }
+
+        if self.upsampling > 1 {
+            if let Some((info, upsampling)) = nonserialized
+                .extra_channel_info
+                .iter()
+                .zip(&self.ec_upsampling)
+                .find(|(info, upsampling)| info.dim_shift * *upsampling < self.upsampling)
```

Comment:
Also need to check if it is > 8

Reply:
Yeah, I'll fix that. Jxl also does some [werid bit shifting](https://gitlab.com/wg1/jpeg-xl/-/blob/main/lib/jxl/frame_header.cc#L209-213) that I don't understand. Do we need to do that as well ?

Reply:
Ah, I'm not sure if dim_shift is a shift count or a multiplier (i.e. if it is 0-1-2 or 1-2-4). I don't think you need to reproduce that bit shifting :)

Reply:
It's << dim_shift and * (ec_)upsampling.
(first edition of the spec had that wrong in one place, I fixed it yesterday in the 2nd edition draft).

So an extra channel of width w should after all upsampling have width w * ec_upsampling << dim_shift. And also (ec_upsampling << dim_shift) >= upsampling (i.e. extra channels cannot be higher resolution than the color channels).